### PR TITLE
fix: CometExec's outputPartitioning might not be same as Spark expects after AQE interferes

### LINKED ISSUE #298: CometExec's outputPartitioning might not be same as Spark expects after AQE interferes
### Describe the bug

Currently, `CometExec` has default `outputPartitioning` implementation that reuses Spark original plan's `outputPartitioning`. In most cases, it is correct. But Spark AQE has special node `AQEShuffleRead` which can possibly change output partitioning for coalescing purpose, e.g.:

```
*(2) ColumnarToRow                                                                                                                                                                                                               
+- CometSort [id#3311, data#3312, day#3313], [data#3312 DESC NULLS FIRST, id#3311 ASC NULLS FIRST]                                                                                                                                     
   +- AQEShuffleRead coalesced                                                                                                                                                                                                         
      +- ShuffleQueryStage 0                                                                                                                                                                                                           
         +- CometColumnarExchange hashpartitioning(data#3312, 5), REPARTITION_BY_COL, CometColumnarShuffle, [plan_id=10211]                                                                                                            
            +- RowToColumnar                                                                                                                                                                                                           
               +- *(1) Project [_1#3304 AS id#3311, _2#3305 AS data#3312, _3#3306 AS day#3313]                                                                                                                                         
                  +- *(1) SerializeFromObject [knownnotnull(assertnotnull(input[0, scala.Tuple3, true]))._1 AS _1#3304, staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, knownnotnull(assertnotnull
(input[0, scala.Tuple3, true]))._2, true, false, true) AS _2#3305, staticinvoke(class org.apache.spark.sql.catalyst.util.DateTimeUtils$, DateType, fromJavaDate, knownnotnull(assertnotnull(input[0, scala.Tuple3, true]))._3, true, fa
lse, true) AS _3#3306]                                                                                                                                                                                                                                      +- Scan[obj#3303]                      
```

It causes many test failures in `org.apache.spark.sql.connector.WriteDistributionAndOrderingSuite` on #250. For example, `CometSort`'s output partitioning is the original plan's `hashpartitioning(data#3312, 5)` of original exchange. But after AQE, `AQEShuffleRead` is added and it modifies output partitioning to `coalescedhashpartitioning(hashpartitioning('data, 5), CoalescedBoundary(0,5))`. Because we replace Spark `SortExec` with `CometSort` before AQE interferes Spark query plan, `CometSort` uses `SortExec`'s output partitioning `hashpartitioning(data#3312, 5)`.


### Steps to reproduce

_No response_

### Expected behavior

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by viirya**
## Which issue does this PR close?

Closes #298.

### CODE REVIEW

File: spark/src/main/scala/org/apache/spark/sql/comet/operators.scala [hunk 1]
```diff
@@ -377,7 +383,8 @@ case class CometProjectExec(
     override val output: Seq[Attribute],
     child: SparkPlan,
     override val serializedPlanOpt: SerializedPlan)
-    extends CometUnaryExec {
+    extends CometUnaryExec
+    with PartitioningPreservingUnaryExecNode {
```
Comment (viirya) [inline spark/src/main/scala/org/apache/spark/sql/comet/operators.scala:387]: `PartitioningPreservingUnaryExecNode` implements `outputPartitioning` for ProjectExec.

File: spark/src/main/scala/org/apache/spark/sql/comet/operators.scala [hunk 2]
```diff
@@ -586,7 +607,8 @@ case class CometHashAggregateExec(
     mode: Option[AggregateMode],
     child: SparkPlan,
     override val serializedPlanOpt: SerializedPlan)
-    extends CometUnaryExec {
+    extends CometUnaryExec
+    with PartitioningPreservingUnaryExecNode {
```
Comment (viirya) [inline spark/src/main/scala/org/apache/spark/sql/comet/operators.scala:611]: PartitioningPreservingUnaryExecNode implements outputPartitioning for `HashAggregateExec`.

File: spark/src/main/scala/org/apache/spark/sql/comet/plans/PartitioningPreservingUnaryExecNode.scala [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+import org.apache.spark.sql.execution.UnaryExecNode
+
+/**
+ * A trait that handles aliases in the `outputExpressions` to produce `outputPartitioning` that
+ * satisfies distribution requirements.
+ *
+ * This is copied from Spark's `PartitioningPreservingUnaryExecNode` because it is only available
+ * in Spark 3.4+. This is a workaround to make it available in Spark 3.2+.
+ */
+trait PartitioningPreservingUnaryExecNode extends UnaryExecNode with AliasAwareOutputExpression {
```
Comment (viirya) [inline spark/src/main/scala/org/apache/spark/sql/comet/plans/PartitioningPreservingUnaryExecNode.scala:35]: Spark 3.4 has a bit more change on `outputPartitioning` of some nodes like `ProjectExec`. Copied it from Spark 3.4.

File: spark/src/main/scala/org/apache/spark/sql/comet/operators.scala [hunk 4]
```diff
@@ -471,6 +488,9 @@ case class CometLocalLimitExec(
     child: SparkPlan,
     override val serializedPlanOpt: SerializedPlan)
     extends CometUnaryExec {
+
+  override def outputPartitioning: Partitioning = child.outputPartitioning
```
Comment (andygrove) [inline spark/src/main/scala/org/apache/spark/sql/comet/operators.scala:492]: Would it make sense to add this `outputPartitioning` to `CometUnaryExec` as the default?

**Review by andygrove — APPROVED** [review]
I'm not an expert on the specific changes but it looks like we are just porting Spark logic over, so LGTM.

**Comment by viirya** [comment]
Thank you @andygrove

**Comment by viirya** [comment]
Merged. Thanks.

**Pull request merged**